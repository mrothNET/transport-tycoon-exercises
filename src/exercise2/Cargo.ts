import { Destination } from "./Destination";
import { Location } from "./Location";

let numCargos = 0;

export class Cargo {
  // tslint:disable-next-line: variable-name
  public readonly cargo_id: number;
  public readonly origin: Location;
  public readonly destination: Destination;

  constructor(origin: Location, destination: Destination) {
    this.cargo_id = ++numCargos;
    this.origin = origin;
    this.destination = destination;
  }
}
