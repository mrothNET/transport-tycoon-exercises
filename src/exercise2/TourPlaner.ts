import { Duration, Time } from "./Time";

export interface TourPlan {
  loading?: Time;
  departure: Time;
  arrival: Time;
  unload?: Time;
  returnDeparture: Time;
  returnArrival: Time;
  cargoAvailableDestination: Time;
}

export class TourPlaner {
  private readonly loadUnload!: boolean;
  private readonly loading!: Duration;
  private readonly travel!: Duration;
  private readonly unloading!: Duration;

  constructor(travel: Duration);
  constructor(loading: Duration, travel: Duration, unloading: Duration);
  constructor(...args: Duration[]) {
    if (args.length === 1) {
      const [travel] = args;
      this.init(false, 0, travel, 0);
    } else if (args.length === 3) {
      const [loading, travel, unloading] = args;
      this.init(true, loading, travel, unloading);
    } else {
      throw new Error("Invalid number of arguments");
    }
  }

  public schedule(startTime: Time): TourPlan {
    const loading = startTime;
    const loadingComplete = loading + this.loading;

    const departure = loadingComplete;
    const arrival = departure + this.travel;

    const unload = arrival;
    const unloadComplete = unload + this.unloading;

    const cargoAvailableDestination = unloadComplete;

    const returnDeparture = unloadComplete;
    const returnArrival = returnDeparture + this.travel;

    if (this.loadUnload) {
      return { loading, departure, arrival, unload, returnDeparture, returnArrival, cargoAvailableDestination };
    } else {
      return { departure, arrival, returnDeparture, returnArrival, cargoAvailableDestination };
    }
  }

  private init(loadUnload: boolean, loading: Duration, travel: Duration, unloading: Duration): void {
    Object.assign(this, { loadUnload, loading, travel, unloading });
  }
}
