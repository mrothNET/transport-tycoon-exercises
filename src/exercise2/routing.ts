import { Location } from "./Location";
import { Duration } from "./Time";

const ROUTING_TABLE: { [origin: string]: { [destination: string]: Duration } } = {
  FACTORY: {
    PORT: 1,
    B: 5,
  },
  PORT: {
    A: 6,
  },
};

export function travelDuration(origin: Location, destination: Location): Duration {
  const routes = ROUTING_TABLE[origin];
  if (typeof routes !== "object") {
    throw new Error(`No routing available at: ${origin}`);
  }

  const duration = routes[destination];
  if (typeof duration !== "number") {
    throw new Error(`No routing available for destination ${destination} at ${origin}`);
  }

  return duration;
}
