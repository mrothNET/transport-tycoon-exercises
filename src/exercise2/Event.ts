import { Cargo } from "./Cargo";
import { Location } from "./Location";
import { Time } from "./Time";

export type EventType = "LOAD" | "DEPART" | "ARRIVE" | "UNLOAD";

export interface Event {
  event: EventType;
  time: Time;
  transport_id: number;
  kind: string;
  location: Location;
  destination?: Location;
  cargo?: Cargo[];
}
