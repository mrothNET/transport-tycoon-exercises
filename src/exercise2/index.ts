export { Time, Duration } from "./Time";
export { Destination, stringToDestination } from "./Destination";
export { TransportTycoon } from "./TransportTycoon";
export { Vehicle } from "./Vehicle";

import { exercise2 } from "./exercise2";
export default exercise2;
