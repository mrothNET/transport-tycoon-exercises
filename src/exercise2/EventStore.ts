import { firstBy } from "thenby";
import { Event } from "./Event";

let numTransportIDs = 0;

export class EventStore {
  private events: Event[] = [];
  private length = 0;

  public push(...events: Event[]): void {
    this.events.push(...events);
  }

  public getAllEvents(): Event[] {
    if (this.events.length !== this.length) {
      this.events.sort(compareEvents);
      this.length = this.events.length;
    }

    return Array.from(this.events);
  }

  public createUniqueTransportID(): number {
    return ++numTransportIDs;
  }
}

const compareEvents = firstBy("time").thenBy(e => eventTypeOrder[(e as Event).event] || 0);

const eventTypeOrder = {
  ARRIVE: -2,
  UNLOAD: -1,
  LOAD: 1,
  DEPART: 2,
};
