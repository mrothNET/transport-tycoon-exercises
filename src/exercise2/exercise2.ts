import { Cargo } from "./Cargo";
import { stringToDestination } from "./Destination";
import { Event } from "./Event";
import { TransportTycoon } from "./TransportTycoon";

export function exercise2(cargoList: string): Event[] {
  const tt = new TransportTycoon();

  cargoList
    .split("")
    .map(stringToDestination)
    .map(destination => new Cargo("FACTORY", destination))
    .forEach(cargo => tt.book(cargo));

  return tt.getAllEvents();
}
