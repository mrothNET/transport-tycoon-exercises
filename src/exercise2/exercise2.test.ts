import { Cargo } from "./Cargo";
import { exercise2 } from "./exercise2";

test('exercise2("B")', () => {
  const events = exercise2("B");
  expect(events.length).toEqual(4);

  const transport_id = events[0].transport_id;
  expect(Number.isInteger(transport_id)).toBe(true);
  expect(transport_id).not.toEqual(0);

  const cargo = events[0].cargo as Cargo[];
  expect(cargo.length).toEqual(1);
  expect(Number.isInteger(cargo[0].cargo_id)).toBe(true);
  expect(cargo[0].cargo_id).not.toEqual(0);
  expect(cargo[0]).toMatchObject({ origin: "FACTORY", destination: "B" });

  expect(events[0]).toMatchObject({ transport_id, time: 0, event: "DEPART", kind: "TRUCK", location: "FACTORY", destination: "B", cargo }); // prettier-ignore
  expect(events[1]).toMatchObject({ transport_id, time: 5, event: "ARRIVE", kind: "TRUCK", location: "B", cargo });
  expect(events[2]).toMatchObject({ transport_id, time: 5, event: "DEPART", kind: "TRUCK", location: "B", destination: "FACTORY" }); // prettier-ignore
  expect(events[3]).toMatchObject({ transport_id, time: 10, event: "ARRIVE", kind: "TRUCK", location: "FACTORY" });

  expect(events[1].destination).toBeUndefined();
  expect(events[3].destination).toBeUndefined();

  expect(events[2].cargo).toBeUndefined();
  expect(events[3].cargo).toBeUndefined();
});

test("updated ship rules", () => {
  const events = exercise2("ABA");

  // @ts-ignore
  const firstCargoId = events[0].cargo[0].cargo_id;
  const secondCargoId = firstCargoId + 1;
  const thirdCargoId = secondCargoId + 1;

  const events1 = events.filter(e => e.cargo && e.cargo[0].cargo_id === firstCargoId);
  expect(events1.map(e => e.event)).toEqual(["DEPART", "ARRIVE", "LOAD", "DEPART", "ARRIVE", "UNLOAD"]);
  expect(events1.map(e => e.time)).toEqual([0, 1, 1, 2, 8, 9]);

  const events3 = events.filter(e => e.cargo && e.cargo[0].cargo_id === thirdCargoId);
  expect(events3.map(e => e.event)).toEqual(["DEPART", "ARRIVE", "LOAD", "DEPART", "ARRIVE", "UNLOAD"]);
  expect(events3.map(e => e.time)).toEqual([2, 3, 15, 16, 22, 23]);
});

test('exercise2("AABABBAB")', () => {
  expect(exercise2("A")).toMatchSnapshot();
  expect(exercise2("B")).toMatchSnapshot();
  expect(exercise2("AA")).toMatchSnapshot();
  expect(exercise2("BB")).toMatchSnapshot();
  expect(exercise2("AB")).toMatchSnapshot();
  expect(exercise2("BA")).toMatchSnapshot();
  expect(exercise2("AABABBAB")).toMatchSnapshot();
});
