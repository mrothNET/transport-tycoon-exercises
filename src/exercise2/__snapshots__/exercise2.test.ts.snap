// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exercise2("AABABBAB") 1`] = `
Array [
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 2,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 2,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 2,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 2,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 2,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "LOAD",
    "kind": "SHIP",
    "location": "PORT",
    "time": 1,
    "transport_id": 3,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 2,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 2,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 2,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "PORT",
    "time": 2,
    "transport_id": 3,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 2,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "A",
    "time": 8,
    "transport_id": 3,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 2,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "UNLOAD",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 3,
  },
  Object {
    "destination": "PORT",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 3,
  },
  Object {
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "PORT",
    "time": 15,
    "transport_id": 3,
  },
]
`;

exports[`exercise2("AABABBAB") 2`] = `
Array [
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 3,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 4,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 3,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 4,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 4,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 10,
    "transport_id": 4,
  },
]
`;

exports[`exercise2("AABABBAB") 3`] = `
Array [
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 4,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 5,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 5,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 6,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 4,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 5,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 5,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 6,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 4,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 5,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "LOAD",
    "kind": "SHIP",
    "location": "PORT",
    "time": 1,
    "transport_id": 7,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 5,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 6,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 5,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 6,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 4,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 5,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "PORT",
    "time": 2,
    "transport_id": 7,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 4,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 5,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "A",
    "time": 8,
    "transport_id": 7,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 4,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 5,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "UNLOAD",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 7,
  },
  Object {
    "destination": "PORT",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 7,
  },
  Object {
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "PORT",
    "time": 15,
    "transport_id": 7,
  },
]
`;

exports[`exercise2("AABABBAB") 4`] = `
Array [
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 6,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 8,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 7,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 9,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 6,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 8,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 7,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 9,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 8,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 9,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 10,
    "transport_id": 8,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 10,
    "transport_id": 9,
  },
]
`;

exports[`exercise2("AABABBAB") 5`] = `
Array [
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 8,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 10,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 9,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 11,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 8,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 10,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 8,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "LOAD",
    "kind": "SHIP",
    "location": "PORT",
    "time": 1,
    "transport_id": 12,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 10,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 10,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 8,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "PORT",
    "time": 2,
    "transport_id": 12,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 9,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 11,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 11,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 8,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "A",
    "time": 8,
    "transport_id": 12,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 8,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "UNLOAD",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 12,
  },
  Object {
    "destination": "PORT",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 12,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 10,
    "transport_id": 11,
  },
  Object {
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "PORT",
    "time": 15,
    "transport_id": 12,
  },
]
`;

exports[`exercise2("AABABBAB") 6`] = `
Array [
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 10,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 13,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 11,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 14,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 11,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 14,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 11,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "LOAD",
    "kind": "SHIP",
    "location": "PORT",
    "time": 1,
    "transport_id": 15,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 14,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 14,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 11,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "PORT",
    "time": 2,
    "transport_id": 15,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 10,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 13,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 5,
    "transport_id": 13,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 11,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "A",
    "time": 8,
    "transport_id": 15,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 11,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "UNLOAD",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 15,
  },
  Object {
    "destination": "PORT",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "A",
    "time": 9,
    "transport_id": 15,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 10,
    "transport_id": 13,
  },
  Object {
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "PORT",
    "time": 15,
    "transport_id": 15,
  },
]
`;

exports[`exercise2("AABABBAB") 7`] = `
Array [
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 12,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 16,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 13,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 0,
    "transport_id": 17,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 12,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 16,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 13,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 17,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 16,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 1,
    "transport_id": 17,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 16,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 17,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 14,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 18,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 15,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 2,
    "transport_id": 19,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 15,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 3,
    "transport_id": 19,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 3,
    "transport_id": 19,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 4,
    "transport_id": 19,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 16,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 4,
    "transport_id": 20,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 14,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 7,
    "transport_id": 18,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 7,
    "transport_id": 18,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 16,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 9,
    "transport_id": 20,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 9,
    "transport_id": 20,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 12,
    "transport_id": 18,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 17,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 12,
    "transport_id": 21,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 14,
    "transport_id": 20,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 18,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "PORT",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 14,
    "transport_id": 22,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 18,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 15,
    "transport_id": 22,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 12,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 13,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 15,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 18,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "LOAD",
    "kind": "SHIP",
    "location": "PORT",
    "time": 15,
    "transport_id": 23,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "PORT",
    "time": 15,
    "transport_id": 22,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 16,
    "transport_id": 22,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 12,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 13,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 15,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 18,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "destination": "A",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "PORT",
    "time": 16,
    "transport_id": 23,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 19,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "destination": "B",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 16,
    "transport_id": 24,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 17,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 17,
    "transport_id": 21,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 17,
    "transport_id": 21,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 19,
        "destination": "B",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "B",
    "time": 21,
    "transport_id": 24,
  },
  Object {
    "destination": "FACTORY",
    "event": "DEPART",
    "kind": "TRUCK",
    "location": "B",
    "time": 21,
    "transport_id": 24,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 22,
    "transport_id": 21,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 12,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 13,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 15,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 18,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "A",
    "time": 22,
    "transport_id": 23,
  },
  Object {
    "cargo": Array [
      Cargo {
        "cargo_id": 12,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 13,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 15,
        "destination": "A",
        "origin": "FACTORY",
      },
      Cargo {
        "cargo_id": 18,
        "destination": "A",
        "origin": "FACTORY",
      },
    ],
    "event": "UNLOAD",
    "kind": "SHIP",
    "location": "A",
    "time": 23,
    "transport_id": 23,
  },
  Object {
    "destination": "PORT",
    "event": "DEPART",
    "kind": "SHIP",
    "location": "A",
    "time": 23,
    "transport_id": 23,
  },
  Object {
    "event": "ARRIVE",
    "kind": "TRUCK",
    "location": "FACTORY",
    "time": 26,
    "transport_id": 24,
  },
  Object {
    "event": "ARRIVE",
    "kind": "SHIP",
    "location": "PORT",
    "time": 29,
    "transport_id": 23,
  },
]
`;
